### PR TITLE
Detect galaxy cluster members in GalSim generated images

### LINKED ISSUE #992: Detect galaxy clusters in GalSim images using the BLISS encoder

### PR DESCRIPTION
**Pull request by kapnadak**
- Wrapped up data generation into a single bash script
- Wrote custom variational distribution
- Wrote metrics for cluster membership (accuracy, precision, recall, F1)
- Modified network to include successive downsampling (to allow for tile sizes larger than 4)
- Set-up a complete pipeline from data generation to training and testing
- Resolves #992 

@gapatron

### CODE REVIEW

**Comment by kapnadak** [comment]
Changes made to main BLISS directory:

1. `image_normalizer.py`: added a check to ensure `self.log_transform_stddevs` is non-empty before iterating
2. `encoder.py`: remove assertion requiring tile length to be 2 or 4, and calculate number of downsampling layers required
3. `convnet.py`: add successive downsampling layers to support tile lengths of greater than 4
4. `catalog.py`: added a few allowed parameters

File: bliss/conf/base_config.yaml [hunk 1]
```diff
@@ -109,8 +109,6 @@ encoder:
             flux_error:
                 _target_: bliss.encoder.metrics.FluxError
                 survey_bands: ${encoder.survey_bands}
-            galaxy_shape_error:
-                _target_: bliss.encoder.metrics.GalaxyShapeError
```
Comment (jeff-regier) [inline bliss/conf/base_config.yaml:113]: Can you override this in your case-study-specific config file, rather than changing the base config file?

File: bliss/encoder/convnet.py [hunk 2]
```diff
@@ -66,7 +66,8 @@ def __init__(self, n_bands, ch_per_band, num_features, double_downsample=True):
             nn.Sequential(*[ConvBlock(64, 64, kernel_size=5, padding=2) for _ in range(4)]),
             ConvBlock(64, 128, stride=2),
             nn.Sequential(*[ConvBlock(128, 128) for _ in range(5)]),
-            ConvBlock(128, num_features, stride=(2 if double_downsample else 1)),  # 4
+            nn.Sequential(*[ConvBlock(128, 128, stride=2) for _ in range(num_downsample)]),
+            ConvBlock(128, num_features, stride=1),  # 4
```
Comment (jeff-regier) [inline bliss/encoder/convnet.py:70]: A potential issue here is that the downsampling doesn't happen until several layers into the network, which sometimes is what you want, but other times isn't. For high-resolution images, you likely want to downsample right away (and to set the kernel size to 5 rather than 3), to reduce memory usage.

File: bliss/encoder/encoder.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 from copy import copy
 from typing import Optional
 
+import numpy as np
```
Comment (jeff-regier) [inline bliss/encoder/encoder.py:4]: can we use torch instead to avoid an import?

File: bliss/conf/base_config.yaml (hunk 1 above)
Comment (kapnadak) [reply bliss/conf/base_config.yaml]: Resolved

File: bliss/encoder/encoder.py [hunk 4]
```diff
@@ -80,14 +81,15 @@ def __init__(
         self.double_detect = double_detect
         self.use_checkerboard = use_checkerboard
 
+        num_downsample = int(np.log2(self.tile_slen)) - 1
```
Comment (jeff-regier) [inline bliss/encoder/encoder.py:84]: Would you assert that `tile_size` is a power of 2 here (or somewhere)?

Also, probably a good idea to round the log of the tile size to the nearest integer before truncating, in case of slight floating point imprecision

File: bliss/encoder/encoder.py [hunk 5]
```diff
@@ -80,14 +81,15 @@ def __init__(
...
+
         ch_per_band = self.image_normalizer.num_channels_per_band()
-        assert tile_slen in {2, 4}, "tile_slen must be 2 or 4"
```
Comment (jeff-regier) [inline bliss/encoder/encoder.py:84]: Would you assert that tile_size is a power of 2 here?

File: bliss/encoder/encoder.py (hunk 3 above)
Comment (kapnadak) [reply bliss/encoder/encoder.py]: Done

File: poetry.lock [hunk 6]
```diff
@@ -592,6 +592,17 @@ files = [
     {file = "cached_property-1.5.2-py2.py3-none-any.whl", hash = "sha256:df4f613cf7ad9a588cc381aaf4a512d26265ecebd5eb9e1ba12f1319eb85a6a0"},
 ]
 
+[[package]]
```
Comment (jeff-regier) [inline poetry.lock:595]: Did you mean to change the poetry.lock file? (Are these changes necessary?)

File: bliss/encoder/encoder.py (hunk 4 above)
Comment (kapnadak) [reply bliss/encoder/encoder.py]: Done

File: bliss/encoder/convnet.py (hunk 2 above)
Comment (kapnadak) [reply bliss/encoder/convnet.py:70]: Pushed all the downsampling layers to the front and increased kernel size to 5. Please let me know if this works!

Comment (jeff-regier) [reply bliss/encoder/convnet.py:70]: Unfortunately not, because this change will break other case studies. Maybe two binary flags is the way to go: one for down sampling here and the other for down sampling right away

Comment (kapnadak) [reply bliss/encoder/convnet.py:70]: Do we want to allow the possibility of downsampling at both places? or just one of the two depending on the flag?

Comment (jeff-regier) [reply bliss/encoder/convnet.py:70]: I'm not really sure, either way seems OK for now

Comment (kapnadak) [reply bliss/encoder/convnet.py]: I added a `downsample_at_front` flag. Seems like the network would have one common backbone, and the flag either puts the downsampling layers at the front or the back, followed by upsampling back to `num_features`. Default value is `false` based on what works for other case studies I think. Added the flag to hydra config as well.

Comment (kapnadak) [reply bliss/encoder/convnet.py]: Resolved for now, things to look into next:
- Upscale number of channels per downsampling layer
- Experiment with attention blocks
- Intersperse successive downsampling layers with C3 blocls
